File: 8_August/Aug_28/rubix.py
import random

class RubiksCube:
    def __init__(self):
        self.faces = {
            'front': [['R' for _ in range(3)] for _ in range(3)],
            'back': [['O' for _ in range(3)] for _ in range(3)],
            'left': [['G' for _ in range(3)] for _ in range(3)],
            'right': [['B' for _ in range(3)] for _ in range(3)],
            'top': [['W' for _ in range(3)] for _ in range(3)],
            'bottom': [['Y' for _ in range(3)] for _ in range(3)]
        }

    def display(self):
        for face in self.faces:
            print(f"{face.capitalize()}:")
            for row in self.faces[face]:
                print(" ".join(row))
            print()

    def rotate_face(self, face, direction):
        if direction == 'clockwise':
            self.faces[face] = [list(row) for row in zip(*self.faces[face][::-1])]
        else:
            self.faces[face] = [list(row) for row in zip(*self.faces[face])][::-1]

    def move(self, face, direction='clockwise'):
        self.rotate_face(face, direction)
        
        if face == 'front':
            self._rotate_adjacent('top', 'right', 'bottom', 'left', 2, direction)
        elif face == 'back':
            self._rotate_adjacent('top', 'left', 'bottom', 'right', 0, direction)
        elif face == 'left':
            self._rotate_adjacent('top', 'front', 'bottom', 'back', 0, direction)
        elif face == 'right':
            self._rotate_adjacent('top', 'back', 'bottom', 'front', 2, direction)
        elif face == 'top':
            self._rotate_adjacent('back', 'right', 'front', 'left', 0, direction)
        elif face == 'bottom':
            self._rotate_adjacent('front', 'right', 'back', 'left', 2, direction)

    def _rotate_adjacent(self, f1, f2, f3, f4, idx, direction):
        temp = [self.faces[f1][i][idx] for i in range(3)]
        if direction == 'clockwise':
            for i in range(3):
                self.faces[f1][i][idx] = self.faces[f4][i][2-idx]
            for i in range(3):
                self.faces[f4][i][2-idx] = self.faces[f3][2-i][2-idx]
            for i in range(3):
                self.faces[f3][i][2-idx] = self.faces[f2][2-i][idx]
            for i in range(3):
                self.faces[f2][i][idx] = temp[i]
        else:
            for i in range(3):
                self.faces[f1][i][idx] = self.faces[f2][i][idx]
            for i in range(3):
                self.faces[f2][i][idx] = self.faces[f3][2-i][2-idx]
            for i in range(3):
                self.faces[f3][i][2-idx] = self.faces[f4][2-i][2-idx]
            for i in range(3):
                self.faces[f4][i][2-idx] = temp[2-i]

    def scramble(self, moves=20):
        faces = list(self.faces.keys())
        directions = ['clockwise', 'counterclockwise']
        for _ in range(moves):
            self.move(random.choice(faces), random.choice(directions))

def main():
    cube = RubiksCube()
    print("Welcome to the Rubik's Cube simulator!")
    print("Commands: 'move <face> [direction]', 'display', 'scramble', 'quit'")
    print("Faces: front, back, left, right, top, bottom")
    print("Directions: clockwise (default), counterclockwise")
    
    while True:
        command = input("Enter command: ").lower().split()
        if not command:
            continue
        
        if command[0] == 'quit':
            break
        elif command[0] == 'display':
            cube.display()
        elif command[0] == 'scramble':
            cube.scramble()
            print("Cube scrambled!")
        elif command[0] == 'move':
            if len(command) < 2:
                print("Please specify a face to move.")
                continue
            face = command[1]
            direction = command[2] if len(command) > 2 else 'clockwise'
            if face not in cube.faces:
                print("Invalid face. Choose from: front, back, left, right, top, bottom")
                continue
            if direction not in ['clockwise', 'counterclockwise']:
                print("Invalid direction. Choose either 'clockwise' or 'counterclockwise'")
                continue
            cube.move(face, direction)
            print(f"Moved {face} {direction}")
        else:
            print("Invalid command. Try 'move', 'display', 'scramble', or 'quit'")

if __name__ == "__main__":
    main()

File: 8_August/Aug_29/rubix.js
const readline = require('readline');

class RubiksCube {
    constructor() {
        this.faces = {
            'front': Array(3).fill().map(() => Array(3).fill('R')),
            'back': Array(3).fill().map(() => Array(3).fill('O')),
            'left': Array(3).fill().map(() => Array(3).fill('G')),
            'right': Array(3).fill().map(() => Array(3).fill('B')),
            'top': Array(3).fill().map(() => Array(3).fill('W')),
            'bottom': Array(3).fill().map(() => Array(3).fill('Y'))
        };
    }

    display() {
        for (const [face, matrix] of Object.entries(this.faces)) {
            console.log(`${face.charAt(0).toUpperCase() + face.slice(1)}:`);
            matrix.forEach(row => console.log(row.join(' ')));
            console.log();
        }
    }

    rotateFace(face, direction) {
        const matrix = this.faces[face];
        if (direction === 'clockwise') {
            this.faces[face] = matrix[0].map((_, i) => matrix.map(row => row[i]).reverse());
        } else {
            this.faces[face] = matrix[0].map((_, i) => matrix.map(row => row[i])).reverse();
        }
    }

    move(face, direction = 'clockwise') {
        this.rotateFace(face, direction);
        
        if (face === 'front') {
            this._rotateAdjacent('top', 'right', 'bottom', 'left', 2, direction);
        } else if (face === 'back') {
            this._rotateAdjacent('top', 'left', 'bottom', 'right', 0, direction);
        } else if (face === 'left') {
            this._rotateAdjacent('top', 'front', 'bottom', 'back', 0, direction);
        } else if (face === 'right') {
            this._rotateAdjacent('top', 'back', 'bottom', 'front', 2, direction);
        } else if (face === 'top') {
            this._rotateAdjacent('back', 'right', 'front', 'left', 0, direction);
        } else if (face === 'bottom') {
            this._rotateAdjacent('front', 'right', 'back', 'left', 2, direction);
        }
    }

    _rotateAdjacent(f1, f2, f3, f4, idx, direction) {
        const temp = this.faces[f1].map(row => row[idx]);
        if (direction === 'clockwise') {
            for (let i = 0; i < 3; i++) {
                this.faces[f1][i][idx] = this.faces[f4][i][2-idx];
                this.faces[f4][i][2-idx] = this.faces[f3][2-i][2-idx];
                this.faces[f3][i][2-idx] = this.faces[f2][2-i][idx];
                this.faces[f2][i][idx] = temp[i];
            }
        } else {
            for (let i = 0; i < 3; i++) {
                this.faces[f1][i][idx] = this.faces[f2][i][idx];
                this.faces[f2][i][idx] = this.faces[f3][2-i][2-idx];
                this.faces[f3][i][2-idx] = this.faces[f4][2-i][2-idx];
                this.faces[f4][i][2-idx] = temp[2-i];
            }
        }
    }

    scramble(moves = 20) {
        const faces = Object.keys(this.faces);
        const directions = ['clockwise', 'counterclockwise'];
        for (let i = 0; i < moves; i++) {
            const randomFace = faces[Math.floor(Math.random() * faces.length)];
            const randomDirection = directions[Math.floor(Math.random() * directions.length)];
            this.move(randomFace, randomDirection);
        }
    }
}

function main() {
    const cube = new RubiksCube();
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    console.log("Welcome to the Rubik's Cube simulator!");
    console.log("Commands: 'move <face> [direction]', 'display', 'scramble', 'quit'");
    console.log("Faces: front, back, left, right, top, bottom");
    console.log("Directions: clockwise (default), counterclockwise");

    function promptUser() {
        rl.question("Enter command: ", (input) => {
            const command = input.toLowerCase().split(' ');

            if (command[0] === 'quit') {
                rl.close();
                return;
            } else if (command[0] === 'display') {
                cube.display();
            } else if (command[0] === 'scramble') {
                cube.scramble();
                console.log("Cube scrambled!");
            } else if (command[0] === 'move') {
                if (command.length < 2) {
                    console.log("Please specify a face to move.");
                } else {
                    const face = command[1];
                    const direction = command[2] || 'clockwise';
                    if (!Object.keys(cube.faces).includes(face)) {
                        console.log("Invalid face. Choose from: front, back, left, right, top, bottom");
                    } else if (!['clockwise', 'counterclockwise'].includes(direction)) {
                        console.log("Invalid direction. Choose either 'clockwise' or 'counterclockwise'");
                    } else {
                        cube.move(face, direction);
                        console.log(`Moved ${face} ${direction}`);
                    }
                }
            } else {
                console.log("Invalid command. Try 'move', 'display', 'scramble', or 'quit'");
            }

            promptUser();
        });
    }

    promptUser();
}

main();